File: ios/Detour/Services/NavigationService.swift
import UIKit
import MapKit
import CoreLocation

enum NavigationApp: String, CaseIterable, Identifiable {
    case appleMaps = "Apple Maps"
    case googleMaps = "Google Maps"
    case waze = "Waze"

    var id: String { rawValue }

    var icon: String {
        switch self {
        case .appleMaps: return "map.fill"
        case .googleMaps: return "globe"
        case .waze: return "car.fill"
        }
    }

    var isInstalled: Bool {
        switch self {
        case .appleMaps: return true
        case .googleMaps:
            return UIApplication.shared.canOpenURL(URL(string: "comgooglemaps://")!)
        case .waze:
            return UIApplication.shared.canOpenURL(URL(string: "waze://")!)
        }
    }

    static var available: [NavigationApp] {
        allCases.filter(\.isInstalled)
    }
}

struct NavigationService {
    static func navigate(
        to poi: POIResult,
        from origin: CLLocationCoordinate2D?,
        originName: String?,
        destination: CLLocationCoordinate2D?,
        destinationName: String?,
        travelMode: String = "DRIVE",
        using app: NavigationApp
    ) {
        navigate(stops: [poi], from: origin, originName: originName, destination: destination, destinationName: destinationName, travelMode: travelMode, using: app)
    }

    static func navigate(
        stops: [POIResult],
        from origin: CLLocationCoordinate2D?,
        originName: String?,
        destination: CLLocationCoordinate2D?,
        destinationName: String?,
        travelMode: String = "DRIVE",
        using app: NavigationApp
    ) {
        AnalyticsService.shared.track("navigation_opened", properties: ["app": app.rawValue, "stopCount": stops.count])
        for poi in stops {
            recordVisit(poi: poi, origin: origin, destination: destination)
        }

        switch app {
        case .appleMaps:
            openAppleMaps(stops: stops, origin: origin, destination: destination, travelMode: travelMode)
        case .googleMaps:
            openGoogleMaps(stops: stops, origin: origin, originName: originName, destination: destination, destinationName: destinationName, travelMode: travelMode)
        case .waze:
            openWaze(poi: stops.first ?? stops[0])
        }
    }

    static func copyAddress(_ poi: POIResult) {
        UIPasteboard.general.string = poi.address
    }

    private static func openAppleMaps(
        stops: [POIResult],
        origin: CLLocationCoordinate2D?,
        destination: CLLocationCoordinate2D?,
        travelMode: String = "DRIVE"
    ) {
        var items: [MKMapItem] = []

        if let origin {
            items.append(MKMapItem(placemark: MKPlacemark(coordinate: origin)))
        } else {
            items.append(.forCurrentLocation())
        }

        for stop in stops {
            let item = MKMapItem(placemark: MKPlacemark(coordinate: stop.coordinate))
            item.name = stop.name
            items.append(item)
        }

        if let destination {
            items.append(MKMapItem(placemark: MKPlacemark(coordinate: destination)))
        }

        MKMapItem.openMaps(
            with: items,
            launchOptions: [MKLaunchOptionsDirectionsModeKey: travelMode == "WALK" ? MKLaunchOptionsDirectionsModeWalking : MKLaunchOptionsDirectionsModeDriving]
        )
    }

    private static func openGoogleMaps(
        stops: [POIResult],
        origin: CLLocationCoordinate2D?,
        originName: String?,
        destination: CLLocationCoordinate2D?,
        destinationName: String?,
        travelMode: String = "DRIVE"
    ) {
        let originStr = originName ?? (origin.map { "\($0.latitude),\($0.longitude)" } ?? "")
        let waypointsStr = stops.map { poi in
            poi.address.isEmpty ? "\(poi.coordinate.latitude),\(poi.coordinate.longitude)" : poi.address
        }.joined(separator: "|")
        let destStr = destinationName ?? (destination.map { "\($0.latitude),\($0.longitude)" } ?? "")

        let gmapMode: String
        switch travelMode {
        case "WALK": gmapMode = "walking"
        case "BICYCLE": gmapMode = "bicycling"
        default: gmapMode = "driving"
        }

        let urlString = "https://www.google.com/maps/dir/?api=1" +
            "&origin=\(originStr)" +
            "&destination=\(destStr)" +
            "&waypoints=\(waypointsStr)" +
            "&travelmode=\(gmapMode)"

        if let encoded = urlString.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed),
           let url = URL(string: encoded) {
            UIApplication.shared.open(url)
        }
    }

    private static func openWaze(poi: POIResult) {
        let urlString = "waze://?ll=\(poi.coordinate.latitude),\(poi.coordinate.longitude)&navigate=yes"
        if let url = URL(string: urlString) {
            UIApplication.shared.open(url)
        }
    }

    private static func recordVisit(
        poi: POIResult,
        origin: CLLocationCoordinate2D?,
        destination: CLLocationCoordinate2D?
    ) {
        var body: [String: Any] = [
            "placeId": poi.placeId,
            "placeName": poi.name,
            "lat": poi.lat,
            "lng": poi.lng,
        ]
        if let o = origin, let d = destination {
            body["originLat"] = o.latitude
            body["originLng"] = o.longitude
            body["destLat"] = d.latitude
            body["destLng"] = d.longitude
        }

        guard let url = URL(string: "\(APIService.baseURL)/api/visit"),
              let jsonData = try? JSONSerialization.data(withJSONObject: body) else { return }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(AnalyticsService.shared.anonymousId, forHTTPHeaderField: "X-Anonymous-Id")
        request.httpBody = jsonData

        Task { _ = try? await URLSession.shared.data(for: request) }
    }
}

File: android/app/src/main/java/com/ahmedkhaled/onroute/service/NavigationService.kt
package com.ahmedkhaled.onroute.service

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import com.ahmedkhaled.onroute.model.POIResult
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.net.URLEncoder

object NavigationService {

    fun openGoogleMaps(
        context: Context,
        poi: POIResult,
        originName: String?,
        destinationName: String?,
        travelMode: String = "DRIVE"
    ) {
        val origin = URLEncoder.encode(originName ?: "", "UTF-8")
        val stop = URLEncoder.encode(poi.address, "UTF-8")
        val dest = URLEncoder.encode(destinationName ?: "", "UTF-8")
        val mode = when (travelMode) {
            "WALK" -> "walking"
            "BICYCLE" -> "bicycling"
            else -> "driving"
        }
        AnalyticsService.track("navigation_opened", mapOf("app" to "Google Maps"))
        recordVisit(poi)
        val url = "https://www.google.com/maps/dir/?api=1&origin=$origin&destination=$dest&waypoints=$stop&travelmode=$mode"
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
        context.startActivity(intent)
    }

    fun openWaze(context: Context, poi: POIResult) {
        AnalyticsService.track("navigation_opened", mapOf("app" to "Waze"))
        recordVisit(poi)
        val url = "waze://?ll=${poi.lat},${poi.lng}&navigate=yes"
        try {
            val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
            context.startActivity(intent)
        } catch (e: Exception) {
            val webUrl = "https://waze.com/ul?ll=${poi.lat},${poi.lng}&navigate=yes"
            context.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(webUrl)))
        }
    }

    fun copyAddress(context: Context, poi: POIResult) {
        val clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        clipboard.setPrimaryClip(ClipData.newPlainText("Address", poi.address))
    }

    private fun recordVisit(poi: POIResult) {
        kotlinx.coroutines.CoroutineScope(kotlinx.coroutines.Dispatchers.IO).launch {
            try {
                val body = org.json.JSONObject().apply {
                    put("placeId", poi.placeId)
                    put("placeName", poi.name)
                    put("lat", poi.lat)
                    put("lng", poi.lng)
                }
                val request = okhttp3.Request.Builder()
                    .url("https://backend-navy-iota.vercel.app/api/visit")
                    .addHeader("X-Anonymous-Id", AnalyticsService.anonymousId)
                    .post(body.toString().toRequestBody("application/json".toMediaType()))
                    .build()
                okhttp3.OkHttpClient().newCall(request).execute().close()
            } catch (_: Exception) { }
        }
    }

    fun sendFeedback(context: Context) {
        val subject = URLEncoder.encode("OnRoute Beta Feedback", "UTF-8")
        val body = URLEncoder.encode(
            "\n\n---\nOnRoute v1.0.0 (Android ${android.os.Build.VERSION.RELEASE})",
            "UTF-8"
        )
        val uri = Uri.parse("mailto:ahmed.khaled.a.mohamed@gmail.com?subject=$subject&body=$body")
        val intent = Intent(Intent.ACTION_SENDTO, uri)
        context.startActivity(intent)
    }
}
